src/irmin-pack/unix/lower.ml

```diff
@@ -80,35 +90,66 @@ module Make_volume (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
        when the store starts. *)
     v ~readonly:false volume_path
 
-  let path t = t.path
-  let control t = t.control
-  let empty_state t = if Option.is_none t.control then `Empty t else `Nonempty t
-
-  let is_empty t =
-    match empty_state t with `Empty _ -> true | `Nonempty _ -> false
+  let path = function Empty { path } -> path | Nonempty { path; _ } -> path
 
-  let nonempty_control (`Nonempty t) =
-    match t.control with
-    | None -> failwith "Nonempty volume has no control"
-    | Some cf -> cf
+  let control = function
+    | Empty _ -> None
+    | Nonempty { control; _ } -> Some control
 
-  let start_offset t = (nonempty_control t).start_offset
-  let end_offset t = (nonempty_control t).end_offset
+  let is_empty = function Empty _ -> true | Nonempty _ -> false
 
-  let contains ~offset t =
-    match empty_state t with
-    | `Empty _ -> false
-    | `Nonempty _ as t ->
+  let contains ~off = function
+    | Empty _ -> false
+    | Nonempty { control; _ } ->
         let open Int63.Syntax in
-        start_offset t <= offset && offset < end_offset t
+        control.start_offset <= off && off < control.end_offset
```

Comment:
Exactly. The start offset should correspond to the suffix start offset when the volume is first filled. For the first volume, this will be 0 + no sparse data. For subsequent volumes, it will be possible to read the commit that is the prefix, but _only_ if you follow the parent of the first commit in the suffix part of the data. A direct read of that commit hash would go to the previous volume.